File: src/transfer/pull.py
"""
全量拉取引擎

从 YoudaoNoteDownload 提取而来，负责递归遍历云端目录并下载全部文件。
YoudaoNoteDownload 仅负责单文件下载，不再包含遍历逻辑。
"""

import logging
import os
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import List, Tuple, TYPE_CHECKING

from src.common import get_script_directory, normalize_sep

if TYPE_CHECKING:
    from src.protocols import DirBrowser
    from src.transfer.download import YoudaoNoteDownload

_DownloadTask = Tuple[str, str, str, int, int]  # (file_id, name, local_dir, mtime, ctime)

DOWNLOAD_WORKERS = 8


class PullEngine:
    """递归拉取云端目录到本地"""

    def __init__(self, api: "DirBrowser", downloader: "YoudaoNoteDownload"):
        self.api = api
        self.downloader = downloader

    def pull_all(self, local_dir: str = None, ydnote_dir: str = None) -> bool:
        """
        全量导出所有笔记

        :param local_dir: 本地目录，为空则使用默认目录
        :param ydnote_dir: 只导出指定的有道云目录，为空则导出全部
        :return: 是否成功
        """
        try:
            if not local_dir:
                local_dir = os.path.join(get_script_directory(), "youdaonote-sync")

            if not os.path.exists(local_dir):
                os.makedirs(local_dir, exist_ok=True)

            root_info = self.api.get_root_dir_info_id()
            root_id = root_info.get('fileEntry', {}).get('id')

            if not root_id:
                logging.error("无法获取根目录 ID")
                return False

            if ydnote_dir:
                dir_info = self.api.get_dir_info_by_id(root_id)
                found = False
                for entry in dir_info.get('entries', []):
                    file_entry = entry.get('fileEntry', {})
                    if file_entry.get('name') == ydnote_dir:
                        root_id = file_entry.get('id')
                        found = True
                        break
                if not found:
                    logging.error(f"未找到指定目录: {ydnote_dir}")
                    return False

            logging.info(f"开始全量导出到: {local_dir}")
            tasks = self._collect_download_tasks(root_id, local_dir)
            logging.info(f"扫描完成: {len(tasks)} 个文件待下载")
            self._execute_downloads(tasks)
            logging.info("全量导出完成!")
            return True

        except (OSError, RuntimeError) as e:
            logging.error(f"全量导出失败: {e}")
            return False

    def _collect_download_tasks(self, dir_id: str, local_dir: str) -> List[_DownloadTask]:
        """递归遍历云端目录树，收集所有文件下载任务（目录立即创建）。"""
        tasks: List[_DownloadTask] = []
        dir_info = self.api.get_dir_info_by_id(dir_id)
        entries = dir_info.get('entries', [])

        for entry in entries:
            file_entry = entry.get('fileEntry', {})
            entry_id = file_entry.get('id', '')
            name = file_entry.get('name', '')
            is_dir = file_entry.get('dir', False)

            if is_dir:
                sub_dir = normalize_sep(os.path.join(local_dir, name))
                os.makedirs(sub_dir, exist_ok=True)
                tasks.extend(self._collect_download_tasks(entry_id, sub_dir))
            else:
                modify_time = file_entry.get('modifyTimeForSort', 0)
                create_time = file_entry.get('createTimeForSort', 0)
                tasks.append((entry_id, name, local_dir, modify_time, create_time))

        return tasks

    def _execute_downloads(self, tasks: List[_DownloadTask]) -> None:
        """并发下载所有文件。"""
        if not tasks:
            return

        workers = min(len(tasks), DOWNLOAD_WORKERS)
        succeeded, failed = 0, 0

        with ThreadPoolExecutor(max_workers=workers) as pool:
            futures = {
                pool.submit(
                    self.downloader.download_file,
                    fid, name, ldir, mtime, ctime,
                ): (fid, name)
                for fid, name, ldir, mtime, ctime in tasks
            }
            for fut in as_completed(futures):
                fid, name = futures[fut]
                try:
                    fut.result()
                    succeeded += 1
                except Exception as e:
                    failed += 1
                    logging.error(f"下载失败: {name} ({fid}) - {e}")

        if failed > 0:
            logging.warning(f"下载统计: 成功={succeeded}, 失败={failed}")

File: ts-src/src/browse/pull.ts
/**
 * Pull engine — recursively download the entire cloud directory tree to local disk.
 *
 * Unlike `sync --pull`, this operates independently of the metadata store.
 * It walks the cloud tree via API and downloads every file, preserving
 * the directory structure.
 *
 * Ported from Python src/transfer/pull.py
 */

import { mkdirSync } from 'node:fs';
import { join } from 'node:path';
import type { YoudaoNoteApi } from '../api/client.js';
import type { DirId, FileId } from '../types/common.js';
import { asDirId, asFileId } from '../types/common.js';
import { downloadFile } from '../execute/download.js';
import { findFolderByPath } from './search.js';

interface DownloadTask {
  fileId: FileId;
  name: string;
  localDir: string;
  modifyTime: number;
}

const CONCURRENT_DOWNLOADS = 8;

export interface PullStats {
  total: number;
  succeeded: number;
  failed: number;
}

/**
 * Recursively download all files from the cloud to `localDir`.
 *
 * @param ydnoteDir  Only export this cloud subdirectory (e.g. "工作/项目"). Empty = root.
 */
export async function pullAll(
  api: YoudaoNoteApi,
  localDir: string,
  ydnoteDir?: string,
): Promise<PullStats> {
  mkdirSync(localDir, { recursive: true });

  let rootId: DirId;
  if (ydnoteDir) {
    const found = await findFolderByPath(api, ydnoteDir);
    if (!found) throw new Error(`Cloud folder not found: ${ydnoteDir}`);
    rootId = found;
  } else {
    rootId = await api.getRootId();
  }

  console.log(`Scanning cloud directory tree…`);
  const tasks = await collectTasks(api, rootId, localDir);
  console.log(`Found ${tasks.length} files to download.`);

  const stats = await executeTasks(api, tasks);
  console.log(
    `Pull complete: ${stats.succeeded} succeeded, ${stats.failed} failed out of ${stats.total}.`,
  );
  return stats;
}

async function collectTasks(
  api: YoudaoNoteApi,
  dirId: DirId,
  localDir: string,
): Promise<DownloadTask[]> {
  const tasks: DownloadTask[] = [];
  const dirInfo = await api.getDirInfoById(dirId);

  for (const entry of dirInfo.entries ?? []) {
    const fe = entry.fileEntry;
    const isDir = Boolean(fe.dir);

    if (isDir) {
      const subDir = join(localDir, fe.name);
      mkdirSync(subDir, { recursive: true });
      const sub = await collectTasks(api, asDirId(fe.id), subDir);
      tasks.push(...sub);
    } else {
      tasks.push({
        fileId: asFileId(fe.id),
        name: fe.name,
        localDir,
        modifyTime: fe.modifyTimeForSort ?? 0,
      });
    }
  }

  return tasks;
}

async function executeTasks(api: YoudaoNoteApi, tasks: DownloadTask[]): Promise<PullStats> {
  let succeeded = 0;
  let failed = 0;

  // Simple concurrency pool
  const queue = [...tasks];

  async function worker(): Promise<void> {
    while (queue.length > 0) {
      const task = queue.shift();
      if (!task) break;
      try {
        const localPath = join(task.localDir, task.name);
        await downloadFile(api, task.fileId, localPath, {
          cloudMtime: task.modifyTime,
        });
        succeeded++;
      } catch (e: unknown) {
        failed++;
        console.error(
          `Download failed: ${task.name} — ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    }
  }

  const workerCount = Math.min(tasks.length, CONCURRENT_DOWNLOADS);
  const workers = Array.from({ length: workerCount }, () => worker());
  await Promise.all(workers);

  return { total: tasks.length, succeeded, failed };
}

/**
 * Download a single folder recursively from the cloud.
 */
export async function downloadFolder(
  api: YoudaoNoteApi,
  dirId: DirId,
  localDir: string,
): Promise<PullStats> {
  mkdirSync(localDir, { recursive: true });

  const tasks = await collectTasks(api, dirId, localDir);
  console.log(`Downloading ${tasks.length} files from folder…`);
  return executeTasks(api, tasks);
}
